### PR TITLE
Feature/09

### PR DESCRIPTION
**Pull request by monkeywithbanana**
Week 9 code samples

### CODE REVIEW

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package academy.pocu.comp2500samples.w09.hashcode;
+
+public class Line {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java]: this is after we learned final keyword for good.. so add final unless u plan to override it later?

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
+    }
+
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof Line)) {
+            return false;
+        }
+
+        if (super.equals(obj)) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java]: shouldn't this be the opposite? 

super.equals == false --> return false?

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Point.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package academy.pocu.comp2500samples.w09.hashcode;
+
+public class Point {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Point.java]: final

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Point.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    public int getY() {
+        return this.y;
+    }
+
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof Point)) {
+            return false;
+        }
+
+        if (super.equals(obj)) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Point.java]: same comment as the other class

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Program.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package academy.pocu.comp2500samples.w09.hashcode;
+
+import java.util.HashSet;
+
+public class Program {
+    public static void main(String[] args) {
+        HashSet<Point> pointHashSet = new HashSet<>();
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Program.java]: variable name change to `points`

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Program.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        Point p1 = new Point(1, 7);
+        Point p2 = new Point(1, 7);
+        Point p3 = new Point(7, 1);
+
+        pointHashSet.add(p1);
+
+        System.out.println(pointHashSet.contains(p2));
+        System.out.println(pointHashSet.contains(p3));
+
+        HashSet<Line> lineHashSet = new HashSet<>();
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/hashcode/Program.java]: lines

File: 09/src/academy/pocu/comp2500samples/w09/magician/Air.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package academy.pocu.comp2500samples.w09.magician;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+public class Air extends Attunement {
+    private final int ELITE_SKILL_COOLDOWN_TIME_SEC = 80;
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Air.java]: ELITE_SKILL_COOLDOWN_SECS

File: 09/src/academy/pocu/comp2500samples/w09/magician/Air.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public void attack() {
+        System.out.println("Electric whip!");
+    }
+
+    public void useEliteSkill(OffsetDateTime dateTime) {
+        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+
+        if (now.compareTo(dateTime.plus(Duration.ofSeconds(ELITE_SKILL_COOLDOWN_TIME_SEC))) > 0) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Air.java]: if expression is quite long.. use final int variables here and there

File: 09/src/academy/pocu/comp2500samples/w09/magician/Earth.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package academy.pocu.comp2500samples.w09.magician;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+public class Earth extends Attunement {
+    private final int ELITE_SKILL_COOLDOWN_TIME_SEC = 100;
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Earth.java]: ELITE_SKILL_COOLDOWN_SECS

File: 09/src/academy/pocu/comp2500samples/w09/magician/Earth.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public void attack() {
+        System.out.println("Stone edge!");
+    }
+
+    public void useEliteSkill(OffsetDateTime dateTime) {
+        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+
+        if (now.compareTo(dateTime.plus(Duration.ofSeconds(ELITE_SKILL_COOLDOWN_TIME_SEC))) > 0) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Earth.java]: same as last file

File: 09/src/academy/pocu/comp2500samples/w09/magician/Fire.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package academy.pocu.comp2500samples.w09.magician;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+public class Fire extends Attunement {
+    private final int ELITE_SKILL_COOLDOWN_TIME_SEC = 90;
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Fire.java]: *_COOLDOWN_SECS

File: 09/src/academy/pocu/comp2500samples/w09/magician/Fire.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public void attack() {
+        System.out.println("Fire~~~ Ball~~~~~!!!");
+    }
+
+    public void useEliteSkill(OffsetDateTime dateTime) {
+        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+
+        if (now.compareTo(dateTime.plus(Duration.ofSeconds(ELITE_SKILL_COOLDOWN_TIME_SEC))) > 0) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Fire.java]: same

File: 09/src/academy/pocu/comp2500samples/w09/magician/Water.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
+package academy.pocu.comp2500samples.w09.magician;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+public class Water extends Attunement {
+    private final int ELITE_SKILL_COOLDOWN_TIME_SEC = 75;
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Water.java]: find all these and fix

File: 09/src/academy/pocu/comp2500samples/w09/magician/Water.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public void attack() {
+        System.out.println("Water gun!");
+    }
+
+    public void useEliteSkill(OffsetDateTime dateTime) {
+        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+
+        if (now.compareTo(dateTime.plus(Duration.ofSeconds(ELITE_SKILL_COOLDOWN_TIME_SEC))) > 0) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/magician/Water.java]: and this as well

File: 09/src/academy/pocu/comp2500samples/w09/objectequality/Line.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+        return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
+    }
+
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof Line)) {
+            return false;
+        }
+
+        if (super.equals(obj)) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/objectequality/Line.java]: double check

File: 09/src/academy/pocu/comp2500samples/w09/objectequality/Point.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    public int getY() {
+        return this.y;
+    }
+
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof Point)) {
+            return false;
+        }
+
+        if (super.equals(obj)) {
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/objectequality/Point.java]: same

File: 09/src/academy/pocu/comp2500samples/w09/objectequality/Program.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+        System.out.println(l1 == l2);
+        System.out.println(l1.equals(l2));
+
+        Line l3 = new Line(new Point(4, 1),
+                new Point(9, 11));
+
+        System.out.println(l1.equals(l3));
+
+        HashSet<Point> hashSet = new HashSet<>();
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/objectequality/Program.java]: points

File: 09/src/academy/pocu/comp2500samples/w09/objectequality/Program.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        System.out.println(p1 == p3);
+        System.out.println(p1.equals(p3));
+
+        Line l1 = new Line(p1, p4);
+        Line l2 = new Line(p1, p4);
+
+        System.out.println(l1 == l2);
+        System.out.println(l1.equals(l2));
+
+        Line l3 = new Line(new Point(4, 1),
```
Comment (popekim) [inline 09/src/academy/pocu/comp2500samples/w09/objectequality/Program.java]: why not create point in separate lines

**Review by popekim — APPROVED** [review]
fix and merge.

File: 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java (hunk 2 above)
Comment (monkeywithbanana) [reply 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java]: hmm? why is this opposite? if super.equals(obj) is true, it means obj is pointing to the same object as 'this'

Comment (popekim) [reply 09/src/academy/pocu/comp2500samples/w09/hashcode/Line.java]: what u have said should be on first line.. of the function

if (obj  == this ) {
  return true;
}

**Pull request merged**